contracts/implementations/nodes/VirtualDevice.sol

```diff
@@ -0,0 +1,234 @@
+//SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.13;
+
+import "../../interfaces/INFT.sol";
+import "../../Eip712/Eip712CheckerInternal.sol";
+import "../../libraries/NodesStorage.sol";
+import "../../libraries/nodes/IntegrationStorage.sol";
+import "../../libraries/nodes/VehicleStorage.sol";
+import "../../libraries/nodes/VirtualDeviceStorage.sol";
+import "../../libraries/MapperStorage.sol";
+
+import "../../shared/Roles.sol";
+import "../../shared/Types.sol";
+
+import "@solidstate/contracts/access/access_control/AccessControlInternal.sol";
+
+/// @title VirtualDevice
+/// @notice Contract that represents the Virtual Device node
+/// @dev It uses the Mapper contract to link Virtual Devices to Vehicles
+contract VirtualDevice is AccessControlInternal {
+    bytes32 private constant MINT_TYPEHASH =
+        keccak256(
+            "MintVirtualDeviceSign(uint256 integrationNode,uint256 vehicleNode,address virtualDeviceAddress)"
+        );
+
+    event VirtualDeviceIdProxySet(address indexed proxy);
+    event VirtualDeviceAttributeAdded(string attribute);
+    event VirtualDeviceAttributeSet(
+        uint256 tokenId,
+        string attribute,
+        string info
+    );
+    event VirtualDeviceNodeMinted(
+        uint256 virtualDeviceNode,
+        uint256 vehicleNode,
```

Comment:
This seems noteworthy, doesn't it?

Reply:
There is a limit of 3 `indexed` per event. Which one do you think would be less worthy?

uint256 virtualDeviceNode
uint256 vehicleNode
address virtualDeviceAddress
address owner

Reply:
Boy that's tough.